Support multiple devices to connect with usb host if usb hub is used

AidenHu opened the pull request:
When USB hub is used, maybe multiple devices are attached, so need to consider about this.

AidenHu committed:
subsys: usb: host: support multiple devices under hub

When hub is used, need to consider about multiple
devices are attached.

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

josuah commented on subsys/usb/host/usbh_device.c:
```diff
@@ -464,11 +466,19 @@ int usbh_device_init(struct usb_device *const udev)
+	k_mutex_lock(&uhs_ctx->mutex, K_FOREVER);
+	SYS_DLIST_FOR_EACH_NODE(&uhs_ctx->udevs, node) {
+		device_count++;
+	}
```
I think you can use `device_count = sys_dlist_len(&uhs_ctx->udevs);` in this case.

[doc](https://docs.zephyrproject.org/latest/doxygen/html/group__doubly-linked-list__apis.html#ga6797f138ad66cfb5ee7918bb12869eac)

josuah reviewed (commented):
Thanks. Some discussion about it below.

tmon-nordic commented on subsys/usb/host/usbh_device.c:
```diff
@@ -464,11 +466,19 @@ int usbh_device_init(struct usb_device *const udev)
+		if (err) {
+			LOG_ERR("Failed to signal bus reset");
+			return err;
+		}
```
What about PORT_RESET hub class feature? Where this should be handled? At high-level Linux solves the issue by having a software-only hub at the top level.

AidenHu committed:
subsys: usb: host: remove root member for usbh_context

add usbh_device_get_root and usbh_device_is_root
function to check root device

Signed-off-by: Aiden Hu <weiwei.hu@nxp.com>

AidenHu commented on subsys/usb/host/usbh_device.c:
```diff
@@ -447,11 +447,42 @@ int usbh_device_set_configuration(struct usb_device *const udev, const uint8_t n
+	else {
+		return 0;
+	}
+}
```
Yes, thanks @josuah. Have updated it.

josuah commented:
Thank you @AidenHu for the modifications. As soon as the CI tests above all pass, and review is complete, this PR can be merged.

![screenshot_2025_1120_0245_57](https://github.com/user-attachments/assets/ab893ed7-bd20-4916-91f1-4569ac3394c2)

It is possible to read more about it [here](https://docs.zephyrproject.org/latest/contribute/guidelines.html#continuous-integration) ("Continuous Integration") if curious.

AidenHu commented on subsys/usb/host/usbh_core.c:
```diff
@@ -46,42 +46,47 @@ static int usbh_event_carrier(const struct device *dev,
+
+	udev = usbh_device_get_root(ctx);
+
+	if (udev) {
```
> Comparing explicitly against `NULL` helps with MISRA compliance.
> 
> See rule 85 here: https://docs.zephyrproject.org/latest/contribute/coding_guidelines/index.html#coding-guidelines
> 
> This is not strictly applied, as I remember, but it can also help keeping the code a bit more homogene.

Thanks, fixed it as the rule.

AidenHu commented on subsys/usb/host/usbh_device.c:
```diff
@@ -447,11 +447,41 @@ int usbh_device_set_configuration(struct usb_device *const udev, const uint8_t n
+
+	if (ctx == NULL) {
+		return NULL;
+	}
```
> I think `ctx` is always guaranteed to be `!= NULL`, and might not need verification.

Thanks, fixed.

GeorgeCGV commented on subsys/usb/host/usbh_core.c:
```diff
@@ -46,42 +46,49 @@ static int usbh_event_carrier(const struct device *dev,
+	k_mutex_unlock(&ctx->mutex);
+
+	err = usbh_device_init(udev);
+	if (ret != 0) {
```
```suggestion
	if (err != 0) {
```
